src/main/java/io/cdap/plugin/salesforce/plugin/source/batch/SalesforceBulkRecordReader.java

```diff
@@ -143,19 +145,24 @@ public void close() throws IOException {
   }
 
   @VisibleForTesting
-  void setupParser(InputStream queryResponseStream) throws IOException {
+  void setupParser(String jobId, String batchId, String[] resultIds) throws IOException, AsyncApiException {
     CSVFormat csvFormat = CSVFormat.DEFAULT.
       withHeader().
       withQuoteMode(QuoteMode.ALL).
       withAllowMissingColumnNames(false);
 
-    csvParser = CSVParser.parse(queryResponseStream, StandardCharsets.UTF_8, csvFormat);
+    for (String resultId: resultIds) {
+      InputStream queryResponseStream = bulkConnection.getQueryResultStream(jobId, batchId, resultId);
```

Comment:
This looks like it doesn't fix the problem, as each InputStream still gets opened here.

Instead, keep track of the result id array along with an index into that array. In nextKeyValue(), if the CSVParser has no more elements and there are more results, open the InputStream for the next result. If there are no more results, the reader is done.

Reply:
thanks, fixed